### PR TITLE
SNOW-574657 in-memory cache impementation

### PR DESCRIPTION
**Pull request by sfc-gh-mkeller**
1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes SNOW-574657

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am modifying authorization mechanisms
   - [ ] I am adding new credentials
   - [x] I am modifying OCSP code: I'm adding the new cache to the OCSP checks
   - [x] I am adding a new dependency: `typing_extensions` for `typing` feature backports

3. Please describe how your code solves the related issue.

   Please write a short description of how your code change solves the related issue.

### CODE REVIEW

**Comment by sfc-gh-kwagner** [comment]
@sfc-gh-mkeller do we need a security review for this since you're modifying the OCSP code?

File: src/snowflake/connector/cache.py [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+
+    def _hit(self, k: K) -> None:
+        """This function gets called when a hit occurs.
+
+        Functions that hit every entry (like values) is not going to count.
+
+        Note that while this function does not interact with lock, but it's only
+        called from contexts where the lock is already held.
+        """
+        del k
```
Comment (sfc-gh-aling) [inline src/snowflake/connector/cache.py:245]: may I ask why `del K` is called here?

**Comment by sfc-gh-mkeller** [comment]
> @sfc-gh-mkeller do we need a security review for this since you're modifying the OCSP code?

I don't think so, I didn't actually change any of the OCSP code. I just save what a function returns into the new cache and return it from cache if it's present. Obviously I can do a review if you think I should do one.

**Comment by sfc-gh-kwagner** [comment]
> > @sfc-gh-mkeller do we need a security review for this since you're modifying the OCSP code?
> 
> I don't think so, I didn't actually change any of the OCSP code. I just save what a function returns into the new cache and return it from cache if it's present. Obviously I can do a review if you think I should do one.

I think we should do one at least for the epic associated with this JIRA. Can you make Harsh the reviewer since this applies to all the ecosystem drivers?

**Comment by sfc-gh-mkeller** [comment]
> > > @sfc-gh-mkeller do we need a security review for this since you're modifying the OCSP code?
> > 
> > 
> > I don't think so, I didn't actually change any of the OCSP code. I just save what a function returns into the new cache and return it from cache if it's present. Obviously I can do a review if you think I should do one.
> 
> I think we should do one at least for the epic associated with this JIRA. Can you make Harsh the reviewer since this applies to all the ecosystem drivers?

Oh, we have done a review for the caching epic. We forgot to record it on Jira though. I'll go and fix this!

**Comment by sfc-gh-kwagner** [comment]
> > > > @sfc-gh-mkeller do we need a security review for this since you're modifying the OCSP code?
> > > 
> > > 
> > > I don't think so, I didn't actually change any of the OCSP code. I just save what a function returns into the new cache and return it from cache if it's present. Obviously I can do a review if you think I should do one.
> > 
> > 
> > I think we should do one at least for the epic associated with this JIRA. Can you make Harsh the reviewer since this applies to all the ecosystem drivers?
> 
> Oh, we have done a review for the caching epic. We forgot to record it on Jira though. I'll go and fix this!

Awesome!

**Review by sfc-gh-kwagner — APPROVED** [review]

**Comment by sfc-gh-mkeller** [comment]
Found it: https://github.com/snowflakedb/security_reviews/pull/1956

File: src/snowflake/connector/cache.py [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+                self.telemetry["size"] = len(self._cache)
+        elif isinstance(other, SFDictCache):
+            to_insert: dict[K, CacheEntry[V]] = {k: v for k, v in other._cache.items()}
+            with self._lock:
+                self._cache.update(to_insert)
+                self.telemetry["size"] = len(self._cache)
+        else:
+            raise TypeError
+
+    def _clear_expired_entries(self) -> None:
```
Comment (sfc-gh-aling) [inline src/snowflake/connector/cache.py:214]: I don't see this method is used?

File: src/snowflake/connector/cache.py [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+    ) -> bool:
+        with self._lock:
+            if key in self._cache.keys():
+                try:
+                    self.__getitem(key, should_record_hits=True)
+                    return True
+                except KeyError:
+                    # Fall through
+                    pass
+        return False
```
Comment (sfc-gh-aling) [inline src/snowflake/connector/cache.py:185]: I think this could be simplified as:
```suggestion
        with self._lock:
            try:
                self.__getitem(key, should_record_hits=True)
                return True
            except KeyError:
                # Fall through
                return False
```

File: src/snowflake/connector/cache.py [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+
+def is_expired(d: datetime.datetime) -> bool:
+    return now() >= d
+
+
+class SFDictCache(Generic[K, V]):
+    """A generic in-memory cache that acts somewhat like a dictionary.
+
+    Notes:
+    - Unlike normal dictionaries keys, values and items return list
```
Comment (sfc-gh-sfan) [inline src/snowflake/connector/cache.py:36]: Maybe it's just me but I find this note not very fluent. How about:

`keys(), values() and items() return list materialized at call time, unlike normal dictionaries that return a view object`

File: src/snowflake/connector/cache.py [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+      materialized at call time, instead of returning a view object.
+    """
+
+    def __init__(
+        self,
+        entry_lifetime: int = 60 * 60 * 24,  # 1 day in seconds
```
Comment (sfc-gh-sfan) [inline src/snowflake/connector/cache.py:42]: Maybe make this value a macro `DAY_IN_SECONDS`?

File: src/snowflake/connector/cache.py [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+        from the other cache.
+        """
+        if isinstance(other, dict):
+            t = now() + self._entry_lifetime
+            to_insert: dict[K, CacheEntry[V]] = {
+                k: CacheEntry(expiry=t, entry=v) for k, v in other.items()
+            }
+            with self._lock:
+                self._cache.update(to_insert)
+                self.telemetry["size"] = len(self._cache)
```
Comment (sfc-gh-sfan) [inline src/snowflake/connector/cache.py:205]: This is shared between two if branches and could be extracted:

```
if ...:
  ...
elif ...:
 ...
else:
  ...

with self._lock:
    self._cache.update(to_insert)
    self.telemetry["size"] = len(self._cache)
```

File: src/snowflake/connector/network.py [hunk 7]
```diff
@@ -907,14 +920,14 @@ def _request_exec_wrapper(
...
-            conn: SnowflakeConnection,
-            timed_out: bool = True
+        self,
+        e: Exception,
+        full_url: str,
+        method: str,
+        retry_timeout: int,
+        retry_count: int,
+        conn: SnowflakeConnection,
+        timed_out: bool = True,
```
Comment (sfc-gh-sfan) [inline src/snowflake/connector/network.py:930]: Do we need a merge gate checking for format? It is a bit weird that this PR contains unrelated format changes.

File: src/snowflake/connector/ocsp_snowflake.py [hunk 8]
```diff
@@ -64,9 +67,30 @@
 from snowflake.connector.telemetry_oob import TelemetryService
 from snowflake.connector.time_util import DecorrelateJitterBackoff
 
+from .cache import SFDictCache
+
+OCSP_CACHE: SFDictCache[
+    tuple[bytes, bytes, int, str],
+    tuple[Exception | None, Certificate, Certificate, CertId, bytes],
+] = SFDictCache(
+    entry_lifetime=60 * 60 * 24,  # A day in seconds
```
Comment (sfc-gh-sfan) [inline src/snowflake/connector/ocsp_snowflake.py:76]: Similarly, use a macro. Maybe the macro could be put into `constants` so it can be shared.

File: src/snowflake/connector/ocsp_snowflake.py [hunk 9]
```diff
@@ -1322,13 +1373,23 @@ def _validate_certificates_sequential(
...
 
         for issuer, subject in cert_data:
-            r = self.validate_by_direct_connection(
-                issuer, subject, telemetry_data, hostname, do_retry=do_retry
-            )
-            results.append(r)
+            cert_id, _ = self.create_ocsp_request(issuer=issuer, subject=subject)
+            cache_key = generate_cache_key(cert_id, subject)
+            cache_hit = OCSP_CACHE.get(cache_key)
+            if cache_hit is None:
```
Comment (sfc-gh-sfan) [inline src/snowflake/connector/ocsp_snowflake.py:1379]: I'm curious if we plan to support a `loader` concept (similar to Guava's loading cache) so if a key is not present in the cache, the cache triggers loader and returns it, and also caches the loaded result for next call.

File: src/snowflake/connector/cache.py [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+
+    @classmethod
+    def from_dict(
+        cls,
+        _dict: dict[K, V],
+        **kw,
+    ) -> Self:
+        cache = cls(**kw)
+        for k, v in _dict.items():
+            cache[k] = v
```
Comment (sfc-gh-aling) [inline src/snowflake/connector/cache.py:57]: I feel we may want to `deepcopy` k and v here to keep an internal copy for the SFDictCache.

otherwise we have to be aware that the SFDictCache is keeping the same reference as the dict derived from

File: src/snowflake/connector/cache.py [hunk 11]
```diff
@@ -0,0 +1,264 @@
...
+
+K = TypeVar("K")
+V = TypeVar("V")
+
+
+def is_expired(d: datetime.datetime) -> bool:
+    return now() >= d
+
+
+class SFDictCache(Generic[K, V]):
```
Comment (sfc-gh-aling) [inline src/snowflake/connector/cache.py:32]: is the cache expected to be internally used only?

my concern is that improper usage pattern leads to memory monotonically increment (as we're only doing lazy deletion here) and after long running, users encounter out of memory error.

or this is not a valid concern?

File: src/snowflake/connector/cache.py (hunk 1 above)
Comment (sfc-gh-mkeller) [reply src/snowflake/connector/cache.py:245]: To get around linters complaining about this argument being unused

File: src/snowflake/connector/ocsp_snowflake.py (hunk 9 above)
Comment (sfc-gh-mkeller) [reply src/snowflake/connector/ocsp_snowflake.py:1379]: This PR only introduces an in-memory cache, so not yet. I'm still flushing out the exact save/load strategies.
I'll keep this is mind for the final strategies though!

File: src/snowflake/connector/cache.py (hunk 11 above)
Comment (sfc-gh-mkeller) [reply src/snowflake/connector/cache.py:32]: Internal only, yes. This exact idea was discussed in the design doc for the project, please take a look at that

**Review by sfc-gh-sfan — APPROVED** [review]
LGTM

File: src/snowflake/connector/cache.py (hunk 10 above)
Comment (sfc-gh-aling) [reply src/snowflake/connector/cache.py:63]: sounds reasonable to me, I don't have strong opinions on deepcopy the objects. As long as we're aware of the "same references" here and know how to deal with it then we are good.

File: src/snowflake/connector/cache.py (hunk 2 above)
Comment (sfc-gh-aling) [reply src/snowflake/connector/cache.py:215]: I see, keeping it should be fine

File: src/snowflake/connector/cache.py (hunk 1 above)
Comment (sfc-gh-aling) [reply src/snowflake/connector/cache.py:246]: I see, I think there should be a better way to ignore the lint error, maybe try adding `# noqa` on the line where the error is reported? https://flake8.pycqa.org/en/3.1.1/user/ignoring-errors.html#in-line-ignoring-errors

I'm not sure what side effect that del K could give us (e.g. reduce ref cnt xx)

File: src/snowflake/connector/cache.py (hunk 11 above)
Comment (sfc-gh-aling) [reply src/snowflake/connector/cache.py:34]: yes, I remember that. we know there's a fixed amount of cache and small size.

File: src/snowflake/connector/cache.py (hunk 1 above)
Comment (sfc-gh-mkeller) [reply src/snowflake/connector/cache.py:246]: It does reduce ref count, but it should only delete the local scope's ref and not the outer scope, but I hear you.
I'll change it for something else!

**Review by sfc-gh-jdu — APPROVED** [review]

**Review by sfc-gh-aling — APPROVED** [review]
LGTM
